tests/test_datastore.py

```diff
@@ -543,5 +549,49 @@ class ChainedDatastoreMemoryPermissionsTestCase(InMemoryDatastorePermissionsTest
     configFile = os.path.join(TESTDIR, "config/basic/chainedDatastore2P.yaml")
 
 
+class ChainedDatastorePerStorePermissionsTests(DatastoreTestsBase, unittest.TestCase):
+    """Test that a chained datastore can control permissions per-datastore
+    even if child datastore would accept."""
+
+    configFile = os.path.join(TESTDIR, "config/basic/chainedDatastorePb.yaml")
+
+    def setUp(self):
+        # Override the working directory before calling the base class
+        self.root = tempfile.mkdtemp(dir=TESTDIR)
+        super().setUp()
+
+    def testPermissions(self):
+        """Test permissions model."""
+        metrics = makeExampleMetrics()
+        datastore = self.makeDatastore()
+
+        sc1 = self.storageClassFactory.getStorageClass("StructuredData")
+        sc2 = self.storageClassFactory.getStorageClass("StructuredDataJson")
+        dimensions = frozenset(("Visit", "PhysicalFilter", "Instrument"))
+        dataId1 = {"visit": 52, "physical_filter": "V", "instrument": "DummyCamComp"}
+        dataId2 = {"visit": 52, "physical_filter": "V", "instrument": "HSC"}
+
+        for datasetTypeName, dataId, sc, accepted in (("metric", dataId1, sc1, (False, True, False)),
+                                                      ("metric2", dataId1, sc1, (False, False, False)),
+                                                      ("metric2", dataId2, sc1, (True, False, False)),
+                                                      ("metric33", dataId2, sc2, (True, True, False)),
+                                                      ("metric2", dataId1, sc2, (False, True, False))):
```

Comment:
Recommend adding a `with self.subTest(...)` here to identify loop iterations (yeah, that doesn't work in pytest yet, but I hope it will someday, and in the meantime it's still useful for debugging when running tests directly with Python).